xds/src/main/java/io/grpc/xds/PriorityLoadBalancer.java

```diff
@@ -0,0 +1,297 @@
+    }
+  }
+
+  private void tryNextPriority(int startIdx) {
+    for (int i = startIdx; i < priorities.size(); i++) {
+      String priority = priorities.get(i);
+      if (!children.containsKey(priority)) {
+        ChildLbState child = new ChildLbState(priority);
+        children.put(priority, child);
+        child.updateResolvedAddresses();
+        return; // Give priority i time to connect.
+      }
+      ChildLbState child = children.get(priority);
+      child.reactivate();
+      if (child.connectivityState.equals(READY) || child.connectivityState.equals(IDLE)) {
+        logger.log(XdsLogLevel.DEBUG, "Shifted to priority {0}", priority);
+        currentPriority = priority;
+        updateOverallState(child.connectivityState, child.picker);
+        for (int j = i + 1; j < priorities.size(); j++) {
+          String p = priorities.get(j);
+          if (children.containsKey(p)) {
+            children.get(p).deactivate();
+          }
+        }
+        return;
+      }
+      if (child.failOverTimer != null && child.failOverTimer.isPending()) {
+        return; // Give priority i time to connect.
+      }
+    }
+    // TODO(zdapeng): Include error details of each priority.
+    logger.log(XdsLogLevel.DEBUG, "All priority failed");
+    updateOverallState(TRANSIENT_FAILURE, new ErrorPicker(Status.UNAVAILABLE));
+  }
+
+  private void updateOverallState(ConnectivityState state, SubchannelPicker picker) {
+    if (!state.equals(currentConnectivityState) || !picker.equals(currentPicker)) {
+      currentConnectivityState = state;
+      currentPicker = picker;
+      helper.updateBalancingState(state, picker);
+    }
+  }
+
+  private final class ChildLbState {
+    final String priority;
+    final ChildHelper childHelper;
+    final GracefulSwitchLoadBalancer lb;
+    // Timer to fail over to the next priority if not connected in 10 sec. Scheduled only once at
+    // child initialization.
+    final ScheduledHandle failOverTimer;
+    // Timer to delay shutdown and deletion of the priority. Scheduled whenever the child is
+    // deactivated.
+    @Nullable ScheduledHandle deletionTimer;
+    @Nullable String policy;
+    ConnectivityState connectivityState = CONNECTING;
+    SubchannelPicker picker = BUFFER_PICKER;
+
+    ChildLbState(final String priority) {
+      this.priority = priority;
+      childHelper = new ChildHelper();
+      lb = new GracefulSwitchLoadBalancer(childHelper);
+
+      class FailOverTask implements Runnable {
+        @Override
+        public void run() {
+          if (deletionTimer != null && deletionTimer.isPending()) {
+            // The child is deactivated.
+            return;
+          }
+          int idx = priorityToIndex.get(priority);
+          updateOverallState(
+              TRANSIENT_FAILURE,
+              new ErrorPicker(
+                  Status.UNAVAILABLE.withDescription(
+                      "Resources for priority " + priority + " not reachable in 10 seconds")));
```

Comment:
"Resources" might be ambiguous. Maybe: "Unable to connect to endpoints with priority ...". Also, the variable `priority` (and keys of `priorityToIndex`) is just a canonical string used to refer to each child policy, it does not reveal any priority information. I believe those names are internal implementation details and have no real meanings (they are created only for easiness of updating existing child policies). So you may want to change misleading usages in this class such as error messages, variable naming, etc.

Reply:
"endpoints" might be too specific, given that priority_lb is not xds-specific.

As for the variable `priority`, do you think the priority index would be better?